File: go/problem28.go
package main

import (
	"fmt"
)

func main() {
	fmt.Println(SpiralDiagonalsSum(5))
}

func printArrayAsMatrix(arr [][]int) {

	for i := 0; i < len(arr); i++ {
		for j := 0; j < len(arr[i]); j++ {
			fmt.Printf("%4d ", arr[i][j])
		}
		fmt.Println()
	}
}

func SpiralDiagonalsSum(limit int) int {

	grid := make([][]int, limit)
	for i := 0; i < limit; i++ {
		grid[i] = make([]int, limit)
	}

	x := limit / 2
	y := x
	num := 1

	for i := 0; i <= limit/2; i++ {

		// Move Right i positions
		for h := 0; h < 2*i+1; h++ {
			grid[x][y] = num
			y = y + 1
			num++
		}
		if num >= limit*limit {
			break
		}

		// Move Down i positions
		for v := 0; v < 2*i+1; v++ {
			grid[x][y] = num
			x = x + 1
			num++
		}

		// Move Left i positions
		for h := 0; h < 2*i+2; h++ {
			grid[x][y] = num
			y = y - 1
			num++
		}

		// Move Up i positions
		for v := 0; v < 2*i+2; v++ {
			grid[x][y] = num
			x = x - 1
			num++
		}

	}

	// fmt.Println("Final Grid is ")
	// printArrayAsMatrix(grid)

	sum := 0

	for p, q := 0, limit-1; p < limit; p, q = p+1, q-1 {
		if p != q {
			sum += grid[p][p] + grid[limit-q-1][q]
		} else {
			sum += grid[p][p]
		}
	}

	return sum
}

File: python/problem28.py
def spiralDiagonalsSum(limit):
    grid = [[0 for x in range(limit)] for y in range(limit)]

    x = limit // 2
    y = x
    num = 1

    for i in range(0, limit // 2 + 1):

        # Move Right i positions
        for h in range(0, 2 * i + 1):
            grid[x][y] = num
            y = y + 1
            num += 1

        if num >= limit * limit:
            break

        # Move Down i positions
        for v in range(0, 2 * i + 1):
            grid[x][y] = num
            x = x + 1
            num += 1

        # Move Left i positions
        for h in range(0, 2 * i + 2):
            grid[x][y] = num
            y = y - 1
            num += 1

        # Move Up i positions
        for v in range(0, 2 * i + 2):
            grid[x][y] = num
            x = x - 1
            num += 1

    sum = 0

    for p in range(0, limit):
        sum = sum + grid[p][p]

    for p in range(limit - 1, -1, -1):
        if p != (limit - p - 1):
            sum = sum + grid[limit - p - 1][p]

    return sum

if __name__=="__main__":
    print(spiralDiagonalsSum(5))
